State machine imp

rohan908 opened the pull request:
State machine ready for review. Missing two major things (other than testing)
1. Needs Coast, DrogueDescent, and MainDescent from rocket state machine that Dashiell said would write
2. Needs speed controller implementation

rohan908 committed:
added refurbished boilerplate and PVKF implementation from rocket code. Now need to fix compile errors and missing imports.

rohan908 committed:
added BLA, fixed protobuff import

rohan908 committed:
updated payload branch with new boilerplate code (PVKF and stuff)

rohan908 committed:
Deleted all references to airbrakes (i think? probably should grep lol) No compile errors! Finally can use new boilerplate state machine imp!

rohan908 committed:
Added all state files. and some implementation. Still need to do the hard part of the righting but need a break ugh

rohan908 committed:
Added Quaternion Utils for converting to rotation matrix and finished JudgeRighting

rohan908 committed:
update submodules

rohan908 committed:
update protobuff submodules

Dash-L committed:
copy some main.cpp stuff over from rocket

Dash-L commented on src/FlightParams.h:
```diff
@@ -0,0 +1,31 @@
+
+constexpr uint8_t MAX_FLAIL_ATTEMPTS = 4;
+
+constexpr uint32_t FLAP_EXTENDED_POS = 1500; //in PWM for fully extended flap
+constexpr uint32_t FLAP_RETRACTED_POS = 1000; //in PWM for fully retracted flap
```
These servos are continuous, so there isn't a PWM value corresponding to extended or retracted. They do have a potentiometer for feedback, which will not be the same as the PWM values, so I would use that. Just be careful not to move the servo too far past either position, since I don't believe there is a mechanical stop at either of these positions, and you could end up in a very weird state.

rohan908 committed:
tumbling state and half of vertical side state done

Dash-L commented on src/states/15-Tumbling.cpp:
```diff
@@ -0,0 +1,32 @@
+
+
+    // talk w colette here about using the gyro... iirc from rbe2002 the gyro has a ton of bias so numerically integrating it to find the current gyro velocity shouldn't work and i need to use the ekf? lowk i dont remember...
+    // or does the gyro measure angular velocity because in that case it works perfect lol and im being dumb... now that i think abouot it, i think im dumb!
```
The gyros measure angular velocity, and the ASM330 (ctx->accel) is in dps and the ICM20948 (ctx->mag) is in rad/s. I think Dan did some testing and it seems the ASM330 has much more noise, so the EKFs have been using the ICM20948 for everything.

Dash-L commented on src/states/16-FullyExtendedDrilling.cpp:
```diff
@@ -0,0 +1,19 @@
+State *FullyExtendedDrilling::loop_impl() {
+    Serial.println("FullyExtendedDrilling looped");
+    
+    //TODO: Write speed controller. Biggest issue rn is knowing what pins give encoder info and then writing the class for that... i have good 2002 code but its been a while
```
I didn't think we had an encoder for the drill, or at least there is no pin for it on the breakout board (see this image of the silkscreen, I can provide actual schematic if you'd like, but not all of the pins are what they say they are...)
![image](https://github.com/user-attachments/assets/a0825b6d-fac4-407a-b4db-f39c53603a58)

Dash-L commented on src/states/5-JudgeRighting.cpp:
```diff
@@ -1,14 +1,42 @@
-    Serial.println("Judge Righting looped");
+    //just in case flaps aren't closed:
+    ctx->vertFlapServo.write(FLAP_RETRACTED_POS);
+    ctx->horzFlapServo.write(FLAP_RETRACTED_POS);
```
See other comment on FlightParams.h, but this won't work. You need some sort of PID here.

Dash-L commented on src/states/6-VerticalSide.cpp:
```diff
@@ -6,9 +6,29 @@ void VerticalSide::initialize_impl() {
-    
-    if (this->currentTime > 5000) {
-        return (State *)new HorizontalSide(this->ctx);
+    ctx->vertFlapServo.write(FLAP_EXTENDED_POS); //money buys pid :) i love axons
```
It may be true that money buys pid, but we don't have that money :( you're gonna have to write your own

rohan908 committed:
vertical side done

Dash-L commented on src/states/8-ExtendingAuger.cpp:
```diff
@@ -0,0 +1,27 @@
+
+    //bc we want the drill to spin at a constant vel
+
+    //TODO: Write speed controller. Biggest issue rn is knowing what pins give encoder info and then writing the class for that... i have good 2002 code but its been a while
```
See comment on FullyExtendedDrilling

rohan908 committed:
horizontal side done

Dash-L commented on src/states/8-ExtendingAuger.cpp:
```diff
@@ -0,0 +1,27 @@
+    // ControlSpeed(ctx->drillServo, DRILL_SPEED);
+
+
+    if (abs((long)(ctx->augerExtServo.read() - AUGER_MAX_EXT_POS)) < AUGER_EXT_POS_BAND // we reached max extension
```
Are you sure `Servo::read` is what you want here? All this tells you is the last PWM value you wrote to the servo with `Servo::write`

rohan908 committed:
fixed state machine bc we dont have access to current sensing... SO LAME. Just combined drilling w extending auger essentially

Dash-L committed:
change telemetry stuff to be payload, not rocket

Dash-L reviewed (changes requested):
I'm about to push a couple of general fixes/updates, and will do the states after that, but in the meantime, here's some feedback.

Dash-L committed:
fix some stuff in FlightParams.h

Dash-L commented:
Also I need to add the thing to the sensors to rotate their axis to match with the rocket when on payload, so this comment is mostly a reminder to myself to do that.

rohan908 committed:
drilling state machine subsection done. TODO: drill speed controller

rohan908 committed:
solid ejection and delivery doors open done

rohan908 committed:
state machine done except for 2 things:add rocket state machine code (coast, drogue, maindescent, ...) 2. Write speed control

rohan908 committed:
removed airbrake servos

rohan908 committed:
Compiled rocket states to up state and down state

rohan908 committed:
pull boilerplate

rohan908 committed:
Added gyroBias computation at Prelaunch and made judge righting fixes:
1) State Debouncer
2) Got rid of quaternion computation to choose side and uses accel exclusively to detect gravity direction

rohan908 committed:
made all abs -> std:::abs

rohan908 committed:
added gyZBias shi to vertical and horizontal side flapping

rohan908 committed:
changed everything from accel to mag aka from ASM to ICM IMU

rohan908 committed:
PID Controller written

rohan908 committed:
Added Watchdog, loop count, flight mode logging, and extra logging

rohan908 committed:
Fixed Rocket Stages to more robust new code transitions

Dash-L committed:
update xbee stuff to match rocket

Dash-L committed:
Merge branch 'stateMachineImp' of https://github.com/WPI-HPRC/payload-2025 into stateMachineImp

rohan908 committed:
PID Fully Implemented

rohan908 committed:
Merge branch 'stateMachineImp' of https://github.com/WPI-HPRC/payload-2025 into stateMachineImp

rohan908 committed:
Add axon logging and control

rohan908 committed:
added TODOs and errorDebugFile Prints

rohan908 committed:
Added Debug Print statements and test states

rohan908 committed:
boilerplate update

AmberDrinkwater49 committed:
Update 12-Recovery.cpp - Note on initalization

Rohan,
I added a comment requesting lines to assert actuators off. This is low priority.
Amber

rohan908 committed:
added debuhg

rohan908 committed:
Merge branch 'stateMachineImp' of https://github.com/WPI-HPRC/payload-2025 into stateMachineImp

kalxed commented on src/FlightParams.h:
```diff
@@ -0,0 +1,50 @@
+
+constexpr float LAUNCH_THRESHHOLD = 5; // [g] upwards acceleration greater than this value = launch detected
+
+constexpr uint32_t BOOST_MAX_TIME = 3000; // [ms] FIXME: put actual value here
```
Fix this? (Remove comment if it's fixed)

kalxed commented on src/FlightParams.h:
```diff
@@ -0,0 +1,50 @@
+constexpr uint32_t BOOST_MAX_TIME = 3000; // [ms] FIXME: put actual value here
+constexpr float BURNOUT_THRESHHOLD = 0.3; // [g] upwards accelerateion smaller than this value = motor burnout detected
+
+constexpr uint32_t COAST_MAX_TIME = 5000; // [ms] FIXME: put actual value here
```
Remove comment if fixed, ask Henry if not he probably knows?

kalxed commented on src/FlightParams.h:
```diff
@@ -0,0 +1,50 @@
+constexpr float MAIN_DESCENT_VELOCITY = 10; // [m/s] magnitude of expected velocity during main descent FIXME: put actual value here
+constexpr float DROGUE_DESCENT_VEL_THRESHHOLD = (std::abs(MAIN_DESCENT_VELOCITY) + std::abs(DROGUE_DESCENT_VELOCITY)) / 2;
+
+constexpr uint32_t MAIN_DESCENT_MAX_TIME = 5000; // [ms] FIXME: put actual value here
```
Definitely update this. Do not want Main Descent ending early.

kalxed commented on src/FlightParams.h:
```diff
@@ -0,0 +1,50 @@
+
+constexpr uint32_t FLAP_EXTENDED_POS = 1500; //in PWM for fully extended flap
+constexpr uint32_t FLAP_RETRACTED_POS = 1000; //in PWM for fully retracted flap
+constexpr uint32_t AUGER_EXT_OUT_SPEED = 1500; //in PWM
```
For speed of servos, which are continuous if you weren't aware, you have to put over 1500, 1500 is the neutral position, over 1500 is some speed the one direction, under 1500 is some speed the other direction

kalxed commented on src/FlightParams.h:
```diff
@@ -0,0 +1,50 @@
+constexpr uint32_t FLAP_RETRACTED_POS = 1000; //in PWM for fully retracted flap
+constexpr uint32_t AUGER_EXT_OUT_SPEED = 1500; //in PWM
+constexpr uint32_t AUGER_EXT_IN_SPEED = 1000; //in PWM
+constexpr uint32_t SOLID_DELIVERY_DOOR_OPEN_POS= 1500; //in PWM
```
Same here

kalxed commented on src/FlightParams.h:
```diff
@@ -0,0 +1,50 @@
+constexpr uint32_t AUGER_EXT_IN_SPEED = 1000; //in PWM
+constexpr uint32_t SOLID_DELIVERY_DOOR_OPEN_POS= 1500; //in PWM
+constexpr uint32_t SOLID_DELIVERY_DOOR_CLOSED_POS = 1000; //in PWM
+constexpr uint32_t SOLID_EJECTION_POS = 1500; //in PWM
```
I bet you wish all of these were "positions" unfortunately, they are not, they are all continuous servos, unless I am severly mistaken, when I tested the current payload, they are all using continuous servos that must be _**timed**_ correctly for the systems to properly work.

kalxed commented on src/FlightParams.h:
```diff
@@ -0,0 +1,50 @@
+
+constexpr uint8_t MAX_FLAIL_ATTEMPTS = 4;
+
+constexpr uint32_t FLAP_EXTENDED_POS = 1500; //in PWM for fully extended flap
```
This is unfortunately untuned, but for your sake, I'm hoping the Axons are in regular servo mode instead of continuous servo mode

kalxed commented on src/config.h:
```diff
@@ -14,17 +18,30 @@
+    #define XBEE_ATTN PD8
 
-    #define LED_PIN PB9
+    #define AXON_VER_OUT_PIN A3
```
These PIN defs don't seem right, they are the same for HOR and VER Axon? Meticulously get the pinouts correctly, they had to be switched on the connector due to the analog in servo values.

kalxed commented on src/config.h:
```diff
@@ -14,17 +18,30 @@
+    #define AXON_HOR_IN_PIN A4
 
-    // Servos for MARS
+    #define AXON_VER_OUT_MIN 1395
```
I'm already confused at all of these values, if these are the tuned AXON values, then move these to FlightParams.h, otherwise I don't know what AXON_POT means or AXON_KP, KI, or KD means so to me this is just confusing

kalxed commented on src/main.cpp:
```diff
@@ -2,70 +2,65 @@
-File file;
-
-#define SD_SPI_SPEED SD_SCK_MHZ(50)
+#define DEBUG_ACCEL
```
Remove these before launch? Probably doesn't matter if they are just printouts

kalxed commented on src/main.cpp:
```diff
@@ -2,70 +2,65 @@
 #endif
-    .gps = new MAX10S(),
+    .gps = MAX10S(),
+    .flightMode = false,
```
👀

kalxed commented on src/main.cpp:
```diff
@@ -89,19 +84,41 @@ void output_byte(uint8_t data, uint pin) {
+void loggingLoop();    // Logging (not called when flightMode is set)
+void occasionalLoop(); // For things like flushing SD card
+
+Looper<FunctionsList<mainLoop, xbeeLoop, loggingLoop, occasionalLoop>,
```
More Dashiell magic I assume, very well done 😄

kalxed commented on src/main.cpp:
```diff
@@ -110,56 +127,226 @@ void setup() {
     SPI.setMOSI(SD_MOSI);
     SPI.begin();
 
+    // while (!Serial)
```
Maybe just remove this altogether to get rid of the risk

kalxed commented on src/main.cpp:
```diff
@@ -110,56 +127,226 @@ void setup() {
 
-    lastTime = millis();
-    lastFlush = millis();
+    IWatchdog.begin(4000000);
```
I don't know what this does but I assume it's ok

kalxed commented on src/main.cpp:
```diff
@@ -110,56 +127,226 @@ void setup() {
-    lastFlush = millis();
+    IWatchdog.begin(4000000);
+
+    ctx.vertFlapServo.init();
```
Init the rest of the servos? Or am I getting the idea that they won't be used? (The fact they are continuous servos just makes it more difficult without any feedback)

kalxed commented on src/main.cpp:
```diff
@@ -110,56 +127,226 @@ void setup() {
+
+void xbeeLoop() { xbee.loop(); }
+
+void EKFLoop() {
```
I know this is probably just too superstitious but does the EKF loop happen to have a max amount of time that it can be running? Is this true for any of the loops? For loops like Xbee or Data cleaning you wouldn't want it to get stuck there. If it's too late it's too late but I would encourage that for "non-essential" aspects of the code.

kalxed commented on src/states/10-SolidEjection.cpp:
```diff
@@ -0,0 +1,18 @@
+
+State* SolidEjection::loop_impl() {
+    Serial.println("Solid Ejection looped");
+    ctx->SolidEjectionServo.write(SOLID_EJECTION_POS);
```
These should be two servo writes, one for the doors and one for the pushing. As well make sure the "POS" is actually treated as a continuous servo. The code should look more like: 
```
if (timepassed< 2000){
    servo.write(this pos)
} else {
    servo.write(1500) //NEUTRAL
}
```

kalxed commented on src/states/11-LiquidDelivery.cpp:
```diff
@@ -7,8 +7,10 @@ void LiquidDelivery::initialize_impl() {
     
-    if (this->currentTime > 5000) {
-        return (State *)new Recovery(this->ctx);
+    ctx->LiquidDeliveryServo.write(LIQUID_DELIVERY_POS);  
```
Once again change to treat this as a continuous servo, because it is, you really should just have to do this for like 10 seconds, also, **liquid delivery doesn't work without the Solenoid being activated**

kalxed commented on src/states/12-Recovery.cpp:
```diff
@@ -1,14 +1,13 @@
 #include "States.h"
 
 void Recovery::initialize_impl() {
+  //TODO: Assert all actuators off here to reduce overheating and unneccessary power draw. - Amber
```
Yeah we'll just use the power control we have for the Deimos board 😆 I don't think this is physically possible with what we have. We can put in neutral position but they will all be held just based on power delivery.

kalxed commented on src/states/12-Recovery.cpp:
```diff
@@ -1,14 +1,13 @@
 }
 
 State *Recovery::loop_impl() {
     Serial.println("Recovery looped");
```
Maybe just remove the print messages as well as they won't help during launch?

kalxed commented on src/states/13-Abort.cpp:
Yeah this probably won't do anything

kalxed commented on src/states/15-Tumbling.cpp:
```diff
@@ -0,0 +1,34 @@
+ 
+State *Tumbling::loop_impl() {
+    //if we came here from VerticalSide or HorizontalSide then we need to close the flaps
+    ctx->vertFlapServo.write(FLAP_RETRACTED_POS, this->currentTime);
```
This will really depend on if the Axons are programmed correctly, just check this and either cry if it's correct or cry if it's not. Otherwise you will use the same sketchy method of timing in microseconds using the microcontroller clock, I don't think there's another way.

kalxed commented on src/states/15-Tumbling.cpp:
```diff
@@ -0,0 +1,34 @@
+    if (gyroData.getLastUpdated() != lastGyroReadTime) {
+        lastGyroReadTime = gyroData.getLastUpdated();
+        // TODO: DO TESTING TO ACCOUNT FOR BIAS
+        if (tumblingDebouncer.update(std::abs(gyroData->gyrZ - ctx->gyZBias) < IS_END_TUMBLING_VEL_THRESHOLD, //TODO: check that axis is correct
```
😵 Yeah really just check this is correct in testing, otherwise you honestly instead should just implement the JudgeRighting after the timer goes up as well, instead of going to SolidDelivery because my guess is that Gyro data won't be accurate enough or that the threshold as of now doesn't make sense in how it's calculated to work correctly. To be fair I'm not entirely sure what the gyro Bias is, but it's the denoising is my guess, so otherwise you really just need to be making sure your debouncer is looking for the difference in values over time, if that's so, you are good to go then. I would still make JudgeRighting go after the timer anyways though.

kalxed commented on src/states/15-Tumbling.cpp:
```diff
@@ -0,0 +1,34 @@
+
+    if (this->currentTime > MAX_TUMBLE_TIME) { //NOTE: if this happens then the gyro is cooked and we've tumbled forever... might as well give up and go to door deploy? :(
+        ctx->errorLogFile.printf("[%d] Tumbling state timed out\n", ::millis());
+        return new SolidDelivery(ctx);
```
Like I said before maybe make this JudgeRighting instead, in case gyro goes haywire. I'll double down on this and say you should treat this state at the minimum or at the worst case like a timer that waits for the payload to stabilize, so therefore you should still *try* to right the payload in the case it doesn't end up stabilizing from gyro.

kalxed commented on src/states/16-FullyExtendedDrilling.cpp:
```diff
@@ -0,0 +1,17 @@
+    Serial.println("FullyExtendedDrilling looped");
+
+    ctx->drillServo.write(DRILL_SPEED);
+    ctx->augerExtServo.write(AUGER_EXT_OUT_SPEED);
```
Check that this isn't a continuous servo 😢, if not, then you should be ok, but also make it so they shut off after a period of time, because it will just fly off the track if it's still just 3D printed.

I would also make it so that the Auger extends for a period of time and then the drill can keep drilling for a while.

kalxed commented on src/states/16-FullyExtendedDrilling.cpp:
```diff
@@ -0,0 +1,17 @@
+    ctx->augerExtServo.write(AUGER_EXT_OUT_SPEED);
+
+    if (this->currentTime > MAX_DRILL_TIME) {
+      return new SolidDelivery(ctx);
```
No current sensor so I would give this like 20 seconds or something? I don't think any testing was done with dirt, or if so, ask them and you can see what they say for timing wise.

kalxed commented on src/states/17-PotentiometerTest.cpp:
```diff
@@ -0,0 +1,19 @@
+    Serial.println("Starting Potentiometer Test State");
+}
+
+State* PotentiometerTest::loop_impl() {
```
👏

kalxed commented on src/states/2-Coast.cpp:
```diff
@@ -1,14 +1,27 @@
+  if (lastBaroReadingTime != baroData.getLastUpdated()) {
+    lastBaroReadingTime = baroData.getLastUpdated();
+
+    ewma.update((baroData->altitude - prevAltitude) * (::millis() - lastBaroReadingTime) / 1000.);
```
Don't know what this is but I assume Dashiell has something to do with it, so I approve 👍

kalxed commented on src/states/2-Coast.cpp:
```diff
@@ -1,14 +1,27 @@
+  }
+
+  //TODO: Check COAST_MAX_TIME
+  if (this->currentTime >= COAST_MAX_TIME) {
```
These I think are less bad to get wrong the one that matters the most is probably MainDescent

kalxed commented on src/states/3-Descent.cpp:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  //TODO: Check MAIN_DESCENT_MAX_TIME
+  if (this->currentTime >= MAIN_DESCENT_MAX_TIME) {
```
This is my only big issue, this number should be large in my opinion, like the maximum possible time it could take for the flight and descent. The payload shouldn't have battery issues, so in this case I don't see why not just ensure it's on the ground before doing any other actions. The hope is that even if it transitioned earlier, Tumbling can still tell it's spinning around in the air perhaps?

kalxed commented on src/states/5-JudgeRighting.cpp:
```diff
@@ -1,14 +1,97 @@
-        return (State *)new VerticalSide(this->ctx);
+    //just in case flaps aren't closed:
+    ctx->vertFlapServo.write(FLAP_RETRACTED_POS, this->currentTime);
+    ctx->horzFlapServo.write(FLAP_RETRACTED_POS, this->currentTime);
```
Check Axon servo function, not sure what mode it is in, may need to be reseated as well? Or if it skips because of the belt then just ensure it's seated correctly before launch. Theoretically. 😭

kalxed commented on src/states/States.h:
```diff
@@ -3,65 +3,147 @@
+        BOTTOM,     // Y- side down
+        LEFT,       // X- side down
+        RIGHT,      // X+ side down
+        UNKNOWN     // No side is clearly down
```
What is the utility of this?

kalxed commented on src/states/5-JudgeRighting.cpp:
```diff
@@ -1,14 +1,97 @@
+            return GroundSide::TOP;
+        }
+    }
+    return GroundSide::UNKNOWN;
```
This can never happen logically?

kalxed commented on src/states/5-JudgeRighting.cpp:
```diff
@@ -1,14 +1,97 @@
+        if (accelY > 0){ //gravity is pointed in the negative y axis (bottom side where drill points)
+            return GroundSide::BOTTOM;
+        }
+        else {
```
This seems like sound logic but the else statements make me skeptical because you don't check the value makes any sense... maybe try to check the value isn't just noise?

kalxed commented on src/states/5-JudgeRighting.cpp:
```diff
@@ -1,14 +1,97 @@
+#endif
+
+        //TODO: Check multistate debouncer works
+        GroundSide debouncedSide = judgeRightingDebouncer.update(currSide, ::millis());
```
Putting faith in this debouncer, I think this should be fine? Just check it's not super noisy would be my advice, I think this logic is one of the few pieces of the payload that we can actually be sure works how we want it coding-wise so I'm sure you can get it right with debug prints.

kalxed commented on src/states/6-VerticalSide.cpp:
```diff
@@ -6,9 +6,33 @@ void VerticalSide::initialize_impl() {
-    
-    if (this->currentTime > 5000) {
-        return (State *)new HorizontalSide(this->ctx);
+    ctx->vertFlapServo.write(FLAP_EXTENDED_POS, this->currentTime);
```
Just checked and this looks like Dashiell magic again, so if the axons are correctly coded and mounted, this should be fine?

kalxed commented on src/states/6-VerticalSide.cpp:
```diff
@@ -6,9 +6,33 @@ void VerticalSide::initialize_impl() {
+        //TODO: Check IS_TUMBLING_VEL_THRESHOLD (includes gyZBias offset works properly)
+        if (isRotatingDebouncer.update(std::abs(gyroData->gyrZ - ctx->gyZBias) > IS_TUMBLING_VEL_THRESHOLD, //TODO: check that axis is correct
+                                        ::millis())) {
+            return new Tumbling(ctx);
```
Very neat, this is even more of a reason to NOT TRANSITION TO SOLID DELIVERY DUE TO TIME CONSTRAINT. Now maybe I take this back because there is a loop otherwise? I think you should make this state do less computation in my opinion and just actuate the servos in two ways and then return to JudgeRighting, updating some ctx variable in the meantime so you can tell how long has passed judgeRighting? I think Dashiell could help if needed.

kalxed commented on src/states/6-VerticalSide.cpp:
```diff
@@ -6,9 +6,33 @@ void VerticalSide::initialize_impl() {
+        }
+    }
+
+    //if we are stuck flapping and we aren't rotating then we are in a bush bruh... BEGIN FLAILING BC ITS SO OVER
```
I love the message but at the same time just try to think of this at worst as a timer before you just try SolidDelivery, try to make sure that it actually works. Also you may just want to simplify this to return to JudgeRighting or something instead? Based on your current code it seems like you could have a loop from Side-->Tumbling-->JudgeRighting-->Side since there is no communication about attempts between these states. Correct me if I'm wrong as well.

kalxed commented on src/states/7-HorizontalSide.cpp:
Same comments really as Vertical Side, choose whether Tumbling is best, and fix the loop in case you do have sensors that go out of whack? There needs to be some global sense of time for how long it has been trying to right itself.

kalxed commented on src/Context.h:
```diff
@@ -1,18 +1,61 @@
+    Servo LiquidDeliveryServo;
+
+    // flags
+    bool inBushTimesFlailed = 0; // if i was the only one looking at this code i would name this: its_so_over = ts pmo 🥀 ... when we get to 4 times flailed then we give up
```
Bool? Do you mean uint32_t?

kalxed commented on src/states/8-ExtendingAuger.cpp:
This is the same as FullyExtendedDrilling? Choose one or the other or make it actually do something different. Otherwise same comments, servos will be timed, not just postional 😢 (I'm sad too)

kalxed commented on src/states/5-JudgeRighting.cpp:
```diff
@@ -1,14 +1,97 @@
+        switch (debouncedSide) {
+            case GroundSide::BOTTOM:
+            // I WOULD LOVE TO HAVE AN IR SENSOR RN  -_-
+                return new ExtendingAuger(ctx);
```
Keep in mind that this is the ONLY PLACE where you transition to Extend Auger, otherwise you are going to SolidDelivery (MAYBE) or just looping around the states if the sensors are going wack. You need to implement a universal tracking of time in this state, and I don't think it's a terrible idea to just do the Auger regardless of what side you are on because it shouldn't have a downside really. So that just means transition to it instead of skipping to Solids.

kalxed commented on src/states/11-LiquidDelivery.cpp:
Need solenoids activated for it to work, pass in context from JudgeRighting or use gyro to figure out which one to activate, or just skip this overall, it will probably end up just damaging the electronics anyways 😢

kalxed commented on src/states/9-SolidDelivery.cpp:
```diff
@@ -0,0 +1,18 @@
+    Serial.println("Solid Delivery looped");
+    
+    ctx->SolidDeliveryDoorServo.write(SOLID_DELIVERY_DOOR_OPEN_POS);
+    ctx->augerExtServo.write(AUGER_EXT_IN_SPEED);
```
Don't do these at the same time create some sense of chronology using IF statements and timing if that's all it takes. Doors are continuous so have fun timing them or just make it so they "open" to the max and don't stop opening or pushing because that at the worst just gets the seeds out of the payload.

kalxed commented on src/states/States.h:
```diff
@@ -3,65 +3,147 @@
+    uint32_t lastBaroReadingTime = 0;
+};
+
+class JudgeRighting : public State {
```
Where is the debouncer defined in here?

kalxed commented on src/states/States.h:
```diff
@@ -3,65 +3,147 @@
     ID_Drill,
     ID_SolidDelivery,
     ID_LiquidDelivery,
+    ID_SolidEjection,
```
File name doesn't line up, is this a git issue or...?

kalxed reviewed (changes requested):
You definitely have to fix the loop that may occur in JudgeRighting between Side-->Tumbling-->JudgeRighting, also not sure about just skipping to SolidDelivery. Otherwise seems like a good job. There are just so many values to tune and many many of the states assume positional servos when they are continuous, which is a big flaw. Also ensure your pins are correct I posted some tricky ones in the chat but I can send you my file to try and help with pinouts. Then the timing I would just refer to people with experience and otherwise just ensure MainDescent timeout is longer than you think the airtime will be so you don't start actuating things in any case. You have plenty of time to spare so I think just use that to your advantage when it comes to switching states, just ensure it's always checking for timing so it can switch to the next state if it's stuck.

cvolf-exe committed:
testing changes

cvolf-exe committed:
payload code good

cvolf-exe committed:
flight code- flown 6-12-25
